Lib/test/test_free_threading/test_heapq.py

```diff
@@ -0,0 +1,271 @@
+        max_heap = self.create_heap(OBJECT_COUNT, HeapKind.MAX)
+        pushpop_items = self.create_random_list(-5_000, 10_000, OBJECT_COUNT)
+
+        def heappushpop_max_func(
+            max_heap: list[int], pushpop_items: list[int]
+        ):
+            for item in pushpop_items:
+                popped_item = heapq.heappushpop_max(max_heap, item)
+                self.assertTrue(popped_item >= item)
+
+        self.run_concurrently(
+            worker_func=heappushpop_max_func,
+            args=(max_heap, pushpop_items),
+            nthreads=NTHREADS,
+        )
+        self.assertEqual(len(max_heap), OBJECT_COUNT)
+        self.assertTrue(self.is_max_heap_property_satisfied(max_heap))
+
+    def test_racing_heapreplace_max(self):
+        max_heap = self.create_heap(OBJECT_COUNT, HeapKind.MAX)
+        replace_items = self.create_random_list(-5_000, 10_000, OBJECT_COUNT)
+
+        def heapreplace_max_func(
+            max_heap: list[int], replace_items: list[int]
+        ):
+            for item in replace_items:
+                popped_item = heapq.heapreplace_max(max_heap, item)
+
+        self.run_concurrently(
+            worker_func=heapreplace_max_func,
+            args=(max_heap, replace_items),
+            nthreads=NTHREADS,
+        )
+        self.assertEqual(len(max_heap), OBJECT_COUNT)
+        self.assertTrue(self.is_max_heap_property_satisfied(max_heap))
+
+    def is_min_heap_property_satisfied(self, heap: list[object]) -> bool:
+        """
+        The value of a parent node should be less than or equal to the
+        values of its children.
+        """
+        return self.is_heap_property_satisfied(heap, HeapKind.MIN)
+
+    def is_max_heap_property_satisfied(self, heap: list[object]) -> bool:
+        """
+        The value of a parent node should be greater than or equal to the
+        values of its children.
+        """
+        return self.is_heap_property_satisfied(heap, HeapKind.MAX)
+
+    @staticmethod
+    def is_heap_property_satisfied(
+        heap: list[object], heap_kind: HeapKind
+    ) -> bool:
+        """
+        Check if the heap property is satisfied.
+        """
+        op = operator.le if heap_kind == HeapKind.MIN else operator.ge
+        # position 0 has no parent
+        for pos in range(1, len(heap)):
+            parent_pos = (pos - 1) >> 1
+            if not op(heap[parent_pos], heap[pos]):
+                return False
+
+        return True
+
+    @staticmethod
+    def is_sorted_ascending(lst: list[object]) -> bool:
+        """
+        Check if the list is sorted in ascending order (non-decreasing).
+        """
+        return all(lst[i - 1] <= lst[i] for i in range(1, len(lst)))
+
+    @staticmethod
+    def is_sorted_descending(lst: list[object]) -> bool:
+        """
+        Check if the list is sorted in descending order (non-increasing).
+        """
+        return all(lst[i - 1] >= lst[i] for i in range(1, len(lst)))
```

Comment:
```suggestion
    def check_invariant(self, heap):
        # Check the heap invariant.
        for pos, item in enumerate(heap):
            if pos: # pos 0 has no parent
                parentpos = (pos-1) >> 1
                self.assertTrue(heap[parentpos] <= item)

    def check_max_invariant(self, heap):
        for pos, item in enumerate(heap[1:], start=1):
            parentpos = (pos - 1) >> 1
            self.assertGreaterEqual(heap[parentpos], item)
```

I don't see a need for the sorted methods too, we can just check the invariant which will simplify this quite a bit? Using the enum for two options also seems over the top, why not just `'min'`/`'max'`?

Reply:
> Using the enum for two options also seems over the top, why not just `'min'`/`'max'`

I believe it might be possible to replace all enum usages with strings, and this could be fine in the test. However, I'm not quite sure that removing just 3 lines of enum definition would improve the code. Personally, I would prefer to keep the enum.

Reply:
> I don't see a need for the sorted methods too, we can just check the invariant which will simplify this quite a bit

The list populated from `heappop()` should be sorted. The heap property is a bit of a weaker check. I could have used `sorted()`, but I didn't want to create a copy.

Reply:
> Reuse our existing checks for tests

Used the checks from `test_heapq`, I think I might refactor the test_heapq as a follow up to make the checks staticmethod.